### PR TITLE
Add message attribute to Exceptions

### PR DESCRIPTION
**Pull request by Laurentww**
## Purpose

Adding a message attribute to PyOptSparse Exceptions which allows for parsing of exceptions.

The motivation behind adding the message attribute is to be able to differentiate between `Error`s during testing. For example, to know whether an error is due to the SNOPT libraries not being available or due to problem option incompatibility.

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing

Raise `Error()` during the evaluation of the problem objective or sensitivities and check that the raised error instance has a `message` attribute.

## Checklist

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [ ] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

File: pyoptsparse/pyOpt_error.py [hunk 1]
```diff
@@ -23,7 +23,11 @@ def __init__(self, message):
                 i += len(word) + 1
         msg += " " * (79 - i) + "|\n" + "+" + "-" * 78 + "+" + "\n"
         print(msg)
-        Exception.__init__(self)
+        self.message = msg
```
Comment (nwu63) [inline pyoptsparse/pyOpt_error.py:26]: Do you want to store the formatted message, or the raw `message` string as the attribute? The latter may be easier to work with if you want to parse it.

File: pyoptsparse/pyOpt_error.py [hunk 2]
```diff
@@ -23,7 +23,11 @@ def __init__(self, message):
...
+        super(Error, self).__init__(message)
```
Comment (nwu63) [inline pyoptsparse/pyOpt_error.py:27]: I would prefer the more modern `super().__init__(message)` here

**Review by nwu63 — COMMENTED** [review]
See comments below. Can you talk about the motivation behind this change, and how you would go about using this?

File: pyoptsparse/pyOpt_error.py (hunk 1 above)
Comment (Laurentww) [reply pyoptsparse/pyOpt_error.py:26]: That makes sense, thanks

File: pyoptsparse/pyOpt_error.py (hunk 2 above)
Comment (Laurentww) [reply pyoptsparse/pyOpt_error.py]: Left it in there for clarity, but I agree

**Comment by Laurentww** [comment]
Updated the PR purpose section

**Comment by nwu63** [comment]
Ah, I see the motivation now. This is something that we had discussed internally also---using just the same `Error` class for everything is probably a bad idea. In the long term, I would actually like to go back to built-in exceptions for most stuff, and maybe even create a few specific exception classes for specific issues (such as error importing the compiled libraries). That way, the error message/catching can be done more programmatically than parsing the message and relying on the message being a certain way.

That being said, I'm fine with this as is, but do watch out for release messages in the future if we decide to overhaul the error system. And if you ever feel like working on that feel free to reach out, we're always happy to take contributions :)

**Review by nwu63 — APPROVED** [review]
Looks good, thanks for adding this to the tests too

**Review by marcomangano — APPROVED** [review]

**Pull request merged**